### PR TITLE
feat(gestalt-di): Compactability with `gestalt-di`

### PR DESCRIPTION
**Pull request by DarkWeird**
## Contains
* new gestalt-related module: `gestalt-di`
It is like gestaltv7, but with gestalt-di specific things.
    * Replaced `Reflections` with `gestalt-di`'s `ClassIndex`
    * Replaced `Reflections`'s `ReflectionUtils` methods with own impl

## Reference:
https://github.com/MovingBlocks/gestalt/pull/87

### CODE REVIEW

File: nui-gestalt-di/build.gradle [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+apply from: "$rootDir/gradle/common.gradle"
+
+repositories {
+    google()
+}
+
+dependencies {
+    api group: 'org.terasology.gestalt', name: 'gestalt-module', version: '7.0.7-SNAPSHOT'
+    api group: 'org.terasology.gestalt', name: 'gestalt-asset-core', version: '7.0.7-SNAPSHOT'
+    api group: 'org.terasology.gestalt', name: 'gestalt-di', version: '7.0.7-SNAPSHOT'
```
Comment (BenjaminAmos) [inline nui-gestalt-di/build.gradle:33]: `gestalt-di` is now a part of gestalt `8.0.0-SNAPSHOT`, so these should be changed.
```suggestion
    api group: 'org.terasology.gestalt', name: 'gestalt-module', version: '8.0.0-SNAPSHOT'
    api group: 'org.terasology.gestalt', name: 'gestalt-asset-core', version: '8.0.0-SNAPSHOT'
    api group: 'org.terasology.gestalt', name: 'gestalt-di', version: '8.0.0-SNAPSHOT'
```

File: nui-gestalt-di/build.gradle [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                        }
+                    }
+                    file("$scanResultDir/META-INF/resources")
+                            .newWriter().withCloseable { writer ->
+                        scanResult.allResources.filter {!it.path.startsWith("META-INF")}.nonClassFilesOnly().each { resource ->
+                            writer.writeLine(resource.getPath())
+                        }
+                    }
+                }
+    }
```
Comment (BenjaminAmos) [inline nui-gestalt-di/build.gradle:109]: These changes will not work as-is, as you did not generate a `META-INF/gestalt-indexes-present` file, which is used to identify which jars have been scanned by gestalt. It's a work-around for a very hard-to-diagnose android quirk. The file can be empty, as its contents is not read. Maybe something like this (untested)?
```groovy
file("$scanResultDir/META-INF/gestalt-indexes-present").createNewFile()
```

**Review by BenjaminAmos — COMMENTED** [review]
I've left a couple of notes about the implementation here. Whilst this approach works for NUI's classes, it does not work for any classes inheriting from NUI classes that are not themselves a part of NUI (e.g. any widget inheriting from `UIWidget` outside of the `nui-*` libraries will never be considered by `gestalt-inject-java`).

To make the transition to NUI 3.0 easier for Destination Sol, `nui-libgdx` currently depends on on `nui-gestalt7` at the moment. If this is planned for Destination Sol, then `nui-libgdx` should be changed to depend on `nui-gestalt-di` instead.

**Comment by pollend** [comment]
@DarkWeird I tried to avoid having to add those annotations but without them the annotation processor was not able to generate the correct index. annotation part of gestalt is decoupled so it should still be possible to use this with gestalt v7. so what do you think we should do @keturn .

**Comment by Cervator** [comment]
FYI: a local snapshot of this PR (merged into `master` today) was built & published to Artifactory as `3di-SNAPSHOT`

I wonder though if we could clean up a bit and organize:

1. Nothing uses `v1.x` anymore so we can delete that branch
2. DS uses `v2.x` but only as a snapshot despite it being released - yet it might as well use v3 I think, so we could probably toss v2 as well? I don't imagine v2 will see any more development on its own
3. Is anything preventing us from releasing v3 right now, then make both TS + DS use that version?
4. After that point we could make a `v3.x` branch then merge this into `master` as a v4 snapshot until ready for DS

**Comment by DarkWeird** [comment]
> @DarkWeird I tried to avoid having to add those annotations but without them the annotation processor was not able to generate the correct index. annotation part of gestalt is decoupled so it should still be possible to use this with gestalt v7. so what do you think we should do @keturn . 

Annotation processor cannot process their dependencies. Only current code.
For dependencies you should implement reflections-like scanner.

But annotation processor can process redirected indexes at current artifact.
Then full version will:
1. Annotation processor for types, which provides by DS(or TS)
2. 'handmade' index for nui at nui-gestalt-di

**Comment by pollend** [comment]
@DarkWeird I think I would be ok with just release another version and the annotations can be decoupled at a later point. rather just go forward with the DI changes then wait for additional changes.

**Review by pollend — APPROVED** [review]

**Comment by DarkWeird** [comment]
@pollend 
I think so.

I remember some thing about nui v4... this is part of nui v4? :)
If yes - feel free for pressing green button after changing PR name/description.

**Pull request merged**